### PR TITLE
Feature/dns resolvconf extension

### PR DESCRIPTION
**Pull request by metelik**
The extension for obtaining the resolver's settings.
Other extensions i.e. adding static DNS settings are on the way. Prefer to review in smaller chunks.

### CODE REVIEW

File: lib/nerves_network/resolvconf.ex [hunk 1]
```diff
@@ -109,12 +128,29 @@ defmodule Nerves.Network.Resolvconf do
     {:reply, :ok, state}
   end
 
+  def handle_call({:set_search, ifname, domains}, _from, state) do
+    state = put_in(state[ifname].search, domains)
+    write_resolvconf(state)
+    {:reply, :ok, state}
```
Comment (adkron) [inline lib/nerves_network/resolvconf.ex]: Since we are always going to return `:ok` maybe we should just change this to `handle_cast`?

File: lib/nerves_network/resolvconf.ex [hunk 2]
```diff
@@ -109,12 +128,29 @@ defmodule Nerves.Network.Resolvconf do
...
     write_resolvconf(state)
     {:reply, :ok, state}
   end
 
+  def handle_call({:settings, ifname}, _from, state) do
+    state  = read_resolvconf(ifname, state)
+    #Elixir.Nerves.Network.Resolvconf: state = %{filename: "/home/motyl/resolv.conf", ifmap: %{"ens33" => %{domain: "eur.gad.schneider-electric.com", ifname: "ens33", ipv4_address: "10.216.251.72", ipv4_broadcast: "", ipv4_gateway: "10.216.251.1", ipv4_subnet_mask: "255.255.255.128", nameservers: ["10.156.118.9", "10.198.90.15"]}}, nameservers: ["10.156.118.9", "10.198.90.15"], search: "eur.gad.schneider-electric.com"}; retval = nil
+    ifmap = state[:ifmap]
+    Logger.debug fn -> "#{__MODULE__}: state = #{inspect state};" end
+    Logger.debug fn -> "#{__MODULE__}: ifmap = #{inspect ifmap}" end
```
Comment (adkron) [inline lib/nerves_network/resolvconf.ex]: Do we need to keep all these debug statements? If so I think it would be cleaner to remove the variables and use `state.ifmap`.

File: lib/nerves_network/resolvconf.ex [hunk 3]
```diff
@@ -168,4 +202,59 @@ defmodule Nerves.Network.Resolvconf do
...
+  defp entry_to_map("domain", value, _map), do: %{search: value}
+  defp entry_to_map("search", value, _map), do: %{search: value}
+  defp entry_to_map("nameserver", value, map) do
+    nameservers = map[:"nameservers"]
+    case nameservers do
+      nil -> %{nameservers: [value]}
+      # We are doing it slow way to preserve the order of occurrence of nameserver entries - up to 3 such entries are supported
+      _ -> %{nameservers: nameservers ++ [value]}
+    end
+  end
```
Comment (adkron) [inline lib/nerves_network/resolvconf.ex:317]: I think the nameservers portion would be much cleaner with a pattern match.

~~~elixir
  defp entry_to_map("nameserver", value, %{nameservers: nil), do: %{nameservers: [value]}
  defp entry_to_map("nameserver", value, %{nameservers: nameservers), do: %{nameservers: nameservers ++ [value]}
~~~

The second could have a guard clause that checks that it is a map. If we need to go that far.

File: lib/nerves_network/resolvconf.ex [hunk 4]
```diff
@@ -109,12 +128,29 @@ defmodule Nerves.Network.Resolvconf do
...
+
   def handle_call({:set_nameservers, ifname, nameservers}, _from, state) do
     state = put_in(state[ifname].nameservers, nameservers)
     write_resolvconf(state)
     {:reply, :ok, state}
   end
 
+  def handle_call({:settings, ifname}, _from, state) do
+    state  = read_resolvconf(ifname, state)
+    #Elixir.Nerves.Network.Resolvconf: state = %{filename: "/home/motyl/resolv.conf", ifmap: %{"ens33" => %{domain: "eur.gad.schneider-electric.com", ifname: "ens33", ipv4_address: "10.216.251.72", ipv4_broadcast: "", ipv4_gateway: "10.216.251.1", ipv4_subnet_mask: "255.255.255.128", nameservers: ["10.156.118.9", "10.198.90.15"]}}, nameservers: ["10.156.118.9", "10.198.90.15"], search: "eur.gad.schneider-electric.com"}; retval = nil
```
Comment (adrigonzo) [inline lib/nerves_network/resolvconf.ex]: if this is a note to self, could you change it into doc format instead ?

**Review by adrigonzo — APPROVED** [review]

**Review by adkron — CHANGES_REQUESTED** [review]
I personally find things hard to follow with all the debug statements left in. I think that, while working, it might be better to use `IO.inspect` since it returns the thing being inspected and can be used inline. It might be helpful to learn about `IEX.pry` for debugging. I also find that pipelines are quite readable and all the variables in the middle take away from the actual logic and algorithms.

File: lib/nerves_network/resolvconf.ex [hunk 5]
```diff
@@ -152,21 +158,29 @@ defmodule Nerves.Network.Resolvconf do
...
-    Logger.debug fn -> "#{__MODULE__}: handle_call: :set_static_domains new_state = #{inspect state}" end
-    write_resolvconf(state)
-    {:reply, :ok, state}
+    #Remove the old static_nameservers from the nameservers superposition
+    ifmap = state.ifmap |> Map.get(ifname, %{})
+
+    current_static_nameservers = Map.get(ifmap, :static_nameservers, [])
+    current_nameservers        = Map.get(ifmap, :nameservers, [])
+
+    nameservers = current_nameservers -- current_static_nameservers
```
Comment (adkron) [inline lib/nerves_network/resolvconf.ex:177]: What if we stored these as MapSets instead? They should be sets and have unique elements. We are also trying to do set operations on them like difference.

Comment (metelik) [reply lib/nerves_network/resolvconf.ex:177]: It actually is a good idea... I was missing the 'set' type in Elixir.. In what version it actually appeared ?

**Review by metelik — COMMENTED** [review]
Re-worked

File: lib/nerves_network/resolvconf.ex [hunk 6]
```diff
@@ -136,36 +222,133 @@ defmodule Nerves.Network.Resolvconf do
...
+  defp entry_to_map("domain", value, _map), do: %{search: value}
+  defp entry_to_map("search", value, _map), do: %{search: value}
+  defp entry_to_map("nameserver", value, map) do
+    nameservers = map[:"nameservers"]
+    case nameservers do
+      nil -> %{nameservers: [value]}
+      # We are doing it slow way to preserve the order of occurrence of nameserver entries - up to 3 such entries are supported
+      _ -> %{nameservers: nameservers ++ [value]}
+    end
+  end
```
Comment (metelik) [inline lib/nerves_network/resolvconf.ex:317]: I tried the map pattern-matching like %{nameservers: nil} and it did not work as expected :/

**Review by metelik — COMMENTED** [review]
did some re-work

File: lib/nerves_network/resolvconf.ex (hunk 5 above)
Comment (michaelkschmidt) [reply lib/nerves_network/resolvconf.ex:177]: Here are the docs https://hexdocs.pm/elixir/MapSet.html#content

It doesn't say when they were added but I think it was after Elixir 1.0

**Review by michaelkschmidt — APPROVED** [review]
I think this is good enough to merge.  Can always cleanup more later with the MapSet stuff

File: lib/nerves_network/resolvconf.ex (hunk 5 above)
Comment (adkron) [reply lib/nerves_network/resolvconf.ex:177]: It was originally called Set, but then they changed the name.

File: lib/nerves_network/resolvconf.ex (hunk 6 above)
Comment (adkron) [reply lib/nerves_network/resolvconf.ex:317]: What? How did it fail?
